beacon_node/network/src/router/gossip_processor.rs

```diff
@@ -0,0 +1,810 @@
+                            );
+                        }
+                    }
+                    // There is no new work event and we are unable to spawn a new worker.
+                    //
+                    // I cannot see any good reason why this would happen.
+                    None => {
+                        warn!(
+                            self.log,
+                            "Unexpected gossip processor condition";
+                            "msg" => "no new work and cannot spawn worker"
+                        );
+                    }
+                    // There is a new work event, but the chain is syncing. Ignore it.
+                    Some(WorkEvent { .. })
+                        if self.network_globals.sync_state.read().is_syncing() =>
+                    {
+                        metrics::inc_counter(&metrics::GOSSIP_PROCESSOR_WORK_EVENTS_IGNORED_TOTAL);
+                        trace!(
+                            self.log,
+                            "Gossip processor skipping work";
+                            "msg" => "chain is syncing"
+                        );
+                    }
+                    // There is a new work event and the chain is not syncing. Process it.
+                    Some(WorkEvent {
+                        message_id,
+                        peer_id,
+                        work,
+                    }) => match work {
+                        Work::Attestation(_) if can_spawn => {
+                            self.spawn_worker(idle_tx.clone(), message_id, peer_id, work)
+                        }
+                        Work::Attestation(attestation) => attestation_queue.push(QueueItem {
+                            message_id,
+                            peer_id,
+                            item: attestation,
+                        }),
+                        Work::Aggregate(_) if can_spawn => {
+                            self.spawn_worker(idle_tx.clone(), message_id, peer_id, work)
+                        }
+                        Work::Aggregate(aggregate) => aggregate_queue.push(QueueItem {
+                            message_id,
+                            peer_id,
+                            item: aggregate,
+                        }),
+                    },
+                }
+
+                metrics::set_gauge(
+                    &metrics::GOSSIP_PROCESSOR_WORKERS_ACTIVE_TOTAL,
+                    self.current_workers as i64,
+                );
+                metrics::set_gauge(
+                    &metrics::GOSSIP_PROCESSOR_UNAGGREGATED_ATTESTATION_QUEUE_TOTAL,
+                    attestation_queue.len() as i64,
+                );
+                metrics::set_gauge(
+                    &metrics::GOSSIP_PROCESSOR_AGGREGATED_ATTESTATION_QUEUE_TOTAL,
+                    aggregate_queue.len() as i64,
+                );
+
+                if initial_aggregate_queue_len != aggregate_queue.len() && aggregate_queue.is_full()
```

Comment:
I don't understand the first half of this condition... is this so that we only print the message when the queue fills up for the first time, and not every subsequent time. I'd almost prefer a `aggregate_queue.len() > initial_aggregate_queue_len` to make the intention clearer

Reply:
Oh yeah, this was a bit of a lazy effort to get de-bounce on logs. I've added something better here: b95fa5b.

Reply:
I almost preferred the simple condition haha, but I guess we could factor the fancy debouncer out and use it elsewhere at some point